File: app/src/main/java/com/example/myapplicatio/fragment/ScheduleFragment.java
package com.example.myapplicatio.fragment;

import android.os.AsyncTask;
import android.support.annotation.Nullable;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.Gravity;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.example.myapplicatio.R;
import com.example.myapplicatio.adapter.ScheduleAdapter;
import com.example.myapplicatio.calendar.OnCalendarClickListener;
import com.example.myapplicatio.calendar.schedule.ScheduleLayout;
import com.example.myapplicatio.calendar.schedule.ScheduleRecyclerView;
import com.example.myapplicatio.common.base.app.BaseFragment;
import com.example.myapplicatio.common.bean.Schedule;
import com.example.myapplicatio.common.listener.OnTaskFinishedListener;
import com.example.myapplicatio.common.util.DeviceUtils;
import com.example.myapplicatio.dialog.SelectDateDialog;
import com.example.myapplicatio.task.schedule.AddScheduleTask;
import com.example.myapplicatio.task.schedule.LoadScheduleTask;
import com.example.myapplicatio.util.ToastUtils;

import java.util.Calendar;
import java.util.List;


public class ScheduleFragment extends BaseFragment implements OnCalendarClickListener, View.OnClickListener,
        OnTaskFinishedListener<List<Schedule>>, SelectDateDialog.OnSelectDateListener {

    private ScheduleLayout slSchedule;
    private ScheduleRecyclerView rvScheduleList;
    private EditText etInputContent;
    private RelativeLayout rLNoTask;

    private ScheduleAdapter mScheduleAdapter;
    private int mCurrentSelectYear, mCurrentSelectMonth, mCurrentSelectDay;
    private long mTime;

    public static ScheduleFragment getInstance() {
        return new ScheduleFragment();
    }

    @Nullable
    @Override
    protected View initContentView(LayoutInflater inflater, @Nullable ViewGroup container) {
        return inflater.inflate(R.layout.fragment_schedule, container, false);
    }

    @Override
    protected void bindView() {
        slSchedule = searchViewById(R.id.slSchedule);
        etInputContent = searchViewById(R.id.etInputContent);
        rLNoTask = searchViewById(R.id.rlNoTask);
        slSchedule.setOnCalendarClickListener(this);
        searchViewById(R.id.ibMainClock).setOnClickListener(this);
        searchViewById(R.id.ibMainOk).setOnClickListener(this);
        searchViewById(R.id.ivMainMenu).setOnClickListener(this);
        searchViewById(R.id.llTitleDate).setOnClickListener(this);
        initScheduleList();
        initBottomInputBar();
    }

    @Override
    protected void initData() {
        super.initData();
        initDate();
    }

    @Override
    protected void bindData() {
        super.bindData();
        resetScheduleList();
    }

    public void resetScheduleList() {
        new LoadScheduleTask(getMActivity(), this, mCurrentSelectYear, mCurrentSelectMonth, mCurrentSelectDay)
                .executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
    }

    private void initDate() {
        Calendar calendar = Calendar.getInstance();
        setCurrentSelectDate(calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH), calendar.get(Calendar.DAY_OF_MONTH));
    }

    @Override
    public void onClickDate(int year, int month, int day) {
        setCurrentSelectDate(year, month, day);
        resetScheduleList();
        getListReminder();
    }

    public void getListReminder() {
        ToastUtils.showShortToast(getContext(), "ss");
        TextView t = (TextView) searchViewById(R.id.tvTitleMonth);
        TextView day = (TextView) searchViewById(R.id.tvTitleDay);
        t.setText(" " + getMonthName(mCurrentSelectMonth));
        day.setText(" " + mCurrentSelectDay);
    }

    @Override
    public void onPageChange(int year, int month, int day) {

    }

    private void initScheduleList() {
        rvScheduleList = slSchedule.getSchedulerRecyclerView();
        LinearLayoutManager manager = new LinearLayoutManager(getMActivity());
        manager.setOrientation(LinearLayoutManager.VERTICAL);
        rvScheduleList.setLayoutManager(manager);
        DefaultItemAnimator itemAnimator = new DefaultItemAnimator();
        itemAnimator.setSupportsChangeAnimations(false);
        rvScheduleList.setItemAnimator(itemAnimator);

        mScheduleAdapter = new ScheduleAdapter(getMActivity(), this);
        rvScheduleList.setAdapter(mScheduleAdapter);

//        var items = ArrayList<MyTextSample>()
//        items.add(MyTextSample("Mine", "Test1"))
//        items.add(MyTextSample("Mine", "Test2"))
//        items.add(MyTextSample("Mine", "Test3"))
//        items.add(MyTextSample("Mine", "Test4"))
//        items.add(MyTextSample("Mine", "Test5"))
//        items.add(MyTextSample("Mine", "Test6"))
//
//        var adapter = MyListAdapter(context!!, items, object : MyListAdapter.ItemClickListener {
//            override fun onItemClick(position: MyTextSample) {
//                Toast.makeText(context, position.detail, Toast.LENGTH_SHORT).show()
//            }
//        })
//        list.layoutManager = LinearLayoutManager(context)
//        list.adapter = adapter
    }

    private void initBottomInputBar() {
        etInputContent.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                etInputContent.setGravity(s.length() == 0 ? Gravity.CENTER : Gravity.CENTER_VERTICAL);
            }
        });
        etInputContent.setOnKeyListener(new View.OnKeyListener() {
            @Override
            public boolean onKey(View v, int keyCode, KeyEvent event) {
                return false;
            }
        });
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.ibMainClock:
                showSelectDateDialog();
                break;
            case R.id.ibMainOk:
                addSchedule();
                break;

            case R.id.ivMainMenu: {
                new SelectDateDialog(getActivity(), new SelectDateDialog.OnSelectDateListener() {
                    @Override
                    public void onSelectDate(int year, int month, int day, long time, int position) {
                        ToastUtils.showShortToast(getContext(), "year:" + year + "." + month + "." + day);
                    }
                }, mCurrentSelectYear, mCurrentSelectMonth, mCurrentSelectDay, Calendar.getInstance().get(Calendar.DAY_OF_WEEK)).show();
            }
            case R.id.llTitleDate: {
                resetMainTitleDate(mCurrentSelectYear, mCurrentSelectMonth, mCurrentSelectDay);
            }
        }
    }

    private void resetMainTitleDate(int mCurrentSelectYear, int mCurrentSelectMonth, int mCurrentSelectDay) {
        TextView day = (TextView) searchViewById(R.id.tvTitleDay);
        TextView month = (TextView) searchViewById(R.id.tvTitleMonth);
        day.setText("" + Calendar.getInstance().get(Calendar.DAY_OF_MONTH));
        month.setText(getMonthName(Calendar.getInstance().get(Calendar.DAY_OF_MONTH)));
    }

    private void showSelectDateDialog() {
        new SelectDateDialog(getMActivity(), this, mCurrentSelectYear, mCurrentSelectMonth, mCurrentSelectDay, slSchedule.getMonthCalendar().getCurrentItem()).show();
    }

    private void closeSoftInput() {
        etInputContent.clearFocus();
        DeviceUtils.closeSoftInput(getMActivity(), etInputContent);
    }

    private void addSchedule() {
        String content = etInputContent.getText().toString();
        if (TextUtils.isEmpty(content)) {
            ToastUtils.showShortToast(getMActivity(), R.string.schedule_input_content_is_no_null);
        } else {
            closeSoftInput();
            Schedule schedule = new Schedule();
            schedule.setTitle(content);
            schedule.setState(0);
            schedule.setTime(mTime);
            schedule.setYear(mCurrentSelectYear);
            schedule.setMonth(mCurrentSelectMonth);
            schedule.setDay(mCurrentSelectDay);
            new AddScheduleTask(getMActivity(), new OnTaskFinishedListener<Schedule>() {
                @Override
                public void onTaskFinished(Schedule data) {
                    if (data != null) {
                        mScheduleAdapter.insertItem(data);
                        etInputContent.getText().clear();
                        rLNoTask.setVisibility(View.GONE);
                        mTime = 0;
                        updateTaskHintUi(mScheduleAdapter.getItemCount() - 2);
                    }
                }
            }, schedule).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
        }
    }

    private void setCurrentSelectDate(int year, int month, int day) {
        mCurrentSelectYear = year;
        mCurrentSelectMonth = month;
        mCurrentSelectDay = day;
    }

    @Override
    public void onTaskFinished(List<Schedule> data) {
        mScheduleAdapter.changeAllData(data);
        rLNoTask.setVisibility(data.size() == 0 ? View.VISIBLE : View.GONE);
        updateTaskHintUi(data.size());
    }

    private void updateTaskHintUi(int size) {
        if (size == 0) {
            slSchedule.removeTaskHint(mCurrentSelectDay);
        } else {
            slSchedule.addTaskHint(mCurrentSelectDay);
        }
    }

    @Override
    public void onSelectDate(final int year, final int month, final int day, long time, int position) {
        slSchedule.getMonthCalendar().setCurrentItem(position);
        slSchedule.postDelayed(new Runnable() {
            @Override
            public void run() {
                slSchedule.getMonthCalendar().getCurrentMonthView().clickThisMonth(year, month, day);
            }
        }, 100);
        mTime = time;
    }

    public int getCurrentCalendarPosition() {
        return slSchedule.getMonthCalendar().getCurrentItem();
    }

    public String getMonthName(int idOfMonth) {
        switch (idOfMonth) {
            case (1): {
                return "Yanuar";
            }
            case (2): {
                return "Febural";
            }
            case (3): {
                return "March";
            }
            case (4): {
                return "April";
            }
            case (5): {
                return "May";
            }
            case (6): {
                return "July";
            }
            case (7): {
                return "June";
            }
            case (8): {
                return "August";
            }
            case (9): {
                return "September";
            }
            case (10): {
                return "October";
            }
            case (11): {
                return "November";
            }
            case (12): {
                return "December";
            }
            default:
                return "Not exists";
        }
    }
}

File: app/src/main/java/com/example/myapplicatio/fragment/ScheduleFragment.kt
package com.example.myapplicatio.fragment

import android.arch.lifecycle.ViewModelProviders
import android.graphics.Color
import android.os.AsyncTask
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.view.*
import android.widget.EditText
import android.widget.RelativeLayout
import android.widget.TextView
import com.example.myapplicatio.R
import com.example.myapplicatio.adapter.ScheduleAdapter
import com.example.myapplicatio.aralash.App
import com.example.myapplicatio.aralash.ReminderViewFactory
import com.example.myapplicatio.aralash.SwipeHelper
import com.example.myapplicatio.aralash.SwipeToDeleteCallback
import com.example.myapplicatio.calendar.OnCalendarClickListener
import com.example.myapplicatio.calendar.RecyclerReminder
import com.example.myapplicatio.calendar.schedule.ScheduleLayout
import com.example.myapplicatio.calendar.schedule.ScheduleRecyclerView
import com.example.myapplicatio.common.base.app.BaseFragment
import com.example.myapplicatio.common.bean.Schedule
import com.example.myapplicatio.common.listener.OnTaskFinishedListener
import com.example.myapplicatio.common.util.DeviceUtils
import com.example.myapplicatio.db.ReminderEntity
import com.example.myapplicatio.db.ReminderModelView
import com.example.myapplicatio.dialog.SelectDateDialog
import com.example.myapplicatio.task.schedule.AddScheduleTask
import com.example.myapplicatio.task.schedule.LoadScheduleTask
import com.example.myapplicatio.util.CalendarUtil
import com.example.myapplicatio.util.ToastUtils
import kotlinx.android.synthetic.main.layout_schedule.*
import java.util.*


class ScheduleFragment : BaseFragment(),
        OnCalendarClickListener,
        View.OnClickListener,
        OnTaskFinishedListener<List<Schedule>>,
        SelectDateDialog.OnSelectDateListener {

    private var slSchedule: ScheduleLayout? = null
    private var rvScheduleList: ScheduleRecyclerView? = null
    private var etInputContent: EditText? = null
    private var rlNoTask: RelativeLayout? = null
    private var mScheduleAdapter: ScheduleAdapter? = null
    private var mCurrentSelectYear: Int = 0
    private var mCurrentSelectMonth: Int = 0
    private var mCurrentSelectDay: Int = 0
    private var mTime: Long = 0

    val currentCalendarPosition: Int
        get() = slSchedule!!.monthCalendar.currentItem

    override fun initContentView(inflater: LayoutInflater, container: ViewGroup?): View? {
        return inflater.inflate(R.layout.fragment_schedule, container, false)
    }


    override fun bindView(){

        val factory = ReminderViewFactory(App.getApplicationContext(context!!))
        val modelView = ViewModelProviders.of(this, factory).get(ReminderModelView::class.java)
        var rv_list = searchViewById<RecyclerView>(R.id.rv_list)
        val allReminder = modelView.getAllReminder()
        rv_list.layoutManager = LinearLayoutManager(context)
        val adapter = RecyclerReminder(context!!, ArrayList(allReminder!!), object : RecyclerReminder.ItemClickListener {
            override fun onItemClick(position: ReminderEntity) {
                ToastUtils.showShortToast(context, position.title)
            }
        })

        var swipe = object : SwipeToDeleteCallback(context!!) {
            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                adapter.removeItem(viewHolder.adapterPosition)
            }
        }

        var itemTouchhelper = ItemTouchHelper(swipe)
        itemTouchhelper.attachToRecyclerView(rv_list)
        rv_list.adapter = adapter

        slSchedule = searchViewById<ScheduleLayout>(R.id.slSchedule)
        rvScheduleList = searchViewById<ScheduleRecyclerView>(R.id.rvScheduleList)
        etInputContent = searchViewById<EditText>(R.id.etInputContent)
        rlNoTask = searchViewById<RelativeLayout>(R.id.rlNoTask)
        slSchedule!!.setOnCalendarClickListener(this)
        searchViewById<View>(R.id.ibMainClock).setOnClickListener(this)
        searchViewById<View>(R.id.ibMainOk).setOnClickListener(this)
        searchViewById<View>(R.id.ivMainMenu).setOnClickListener(this)
        searchViewById<View>(R.id.llTitleDate).setOnClickListener(this)
        initScheduleList()
        initBottomInputBar()
    }

    override fun initData() {
        super.initData()
        initDate()
    }

    override fun bindData() {
        super.bindData()
        resetScheduleList()
    }

    fun resetScheduleList() {
        LoadScheduleTask(mActivity, this, mCurrentSelectYear, mCurrentSelectMonth, mCurrentSelectDay)
                .executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR)
    }

    private fun initDate() {
        val calendar = Calendar.getInstance()
        setCurrentSelectDate(calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH), calendar.get(Calendar.DAY_OF_MONTH))
    }

    override fun onClickDate(year: Int, month: Int, day: Int) {
        setCurrentSelectDate(year, month, day)
        resetScheduleList()
        getListReminder()
    }

    fun getListReminder() {
        ToastUtils.showShortToast(context, "ss")
        val t = searchViewById<View>(R.id.tvTitleMonth) as TextView
        val day = searchViewById<View>(R.id.tvTitleDay) as TextView
        t.text = CalendarUtil.MONTH(mCurrentSelectMonth)
        day.text = " $mCurrentSelectDay"
    }

    override fun onPageChange(year: Int, month: Int, day: Int) {

    }

    private fun initScheduleList() {
        val manager = LinearLayoutManager(mActivity)
        manager.orientation = LinearLayoutManager.VERTICAL
        rvScheduleList!!.layoutManager = manager
        val itemAnimator = DefaultItemAnimator()
        itemAnimator.supportsChangeAnimations = false
        rvScheduleList!!.itemAnimator = itemAnimator

        mScheduleAdapter = ScheduleAdapter(mActivity, this)
        rvScheduleList!!.adapter = mScheduleAdapter

        //        var items = ArrayList<MyTextSample>()
        //        items.add(MyTextSample("Mine", "Test1"))
        //        items.add(MyTextSample("Mine", "Test2"))
        //        items.add(MyTextSample("Mine", "Test3"))
        //        items.add(MyTextSample("Mine", "Test4"))
        //        items.add(MyTextSample("Mine", "Test5"))
        //        items.add(MyTextSample("Mine", "Test6"))
        //
        //        var adapter = MyListAdapter(context!!, items, object : MyListAdapter.ItemClickListener {
        //            override fun onItemClick(position: MyTextSample) {
        //                Toast.makeText(context, position.detail, Toast.LENGTH_SHORT).show()
        //            }
        //        })
//                list.layoutManager = LinearLayoutManager(context)
//                list.adapter = adapter
    }

    private fun initBottomInputBar() {
        etInputContent!!.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {

            }

            override fun afterTextChanged(s: Editable) {
                etInputContent!!.gravity = if (s.length == 0) Gravity.CENTER else Gravity.CENTER_VERTICAL
            }
        })
        etInputContent!!.setOnKeyListener { v, keyCode, event -> false }
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.ibMainClock -> showSelectDateDialog()
            R.id.ibMainOk -> addSchedule()

            R.id.ivMainMenu -> {
                run {
                    SelectDateDialog(activity, SelectDateDialog.OnSelectDateListener { year, month, day, time, position -> ToastUtils.showShortToast(context, "year:$year.$month.$day") }, mCurrentSelectYear, mCurrentSelectMonth, mCurrentSelectDay, Calendar.getInstance().get(Calendar.DAY_OF_WEEK)).show()
                }
                run { resetMainTitleDate(mCurrentSelectYear, mCurrentSelectMonth, mCurrentSelectDay) }
            }
            R.id.llTitleDate -> {
                resetMainTitleDate(mCurrentSelectYear, mCurrentSelectMonth, mCurrentSelectDay)
            }
        }
    }

    private fun resetMainTitleDate(mCurrentSelectYear: Int, mCurrentSelectMonth: Int, mCurrentSelectDay: Int) {
        val day = searchViewById<View>(R.id.tvTitleDay) as TextView
        val month = searchViewById<View>(R.id.tvTitleMonth) as TextView
        day.text = CalendarUtil.DAY()
        month.text = CalendarUtil.MONTH()
    }

    private fun showSelectDateDialog() {
        SelectDateDialog(mActivity, this, mCurrentSelectYear, mCurrentSelectMonth, mCurrentSelectDay, slSchedule!!.monthCalendar.currentItem).show()
    }

    private fun closeSoftInput() {
        etInputContent!!.clearFocus()
        DeviceUtils.closeSoftInput(mActivity!!, etInputContent!!)
    }

    private fun addSchedule() {
        val content = etInputContent!!.text.toString()
        if (TextUtils.isEmpty(content)) {
            ToastUtils.showShortToast(mActivity, R.string.schedule_input_content_is_no_null)
        } else {
            closeSoftInput()
            val schedule = Schedule()
            schedule.title = content
            schedule.state = 0
            schedule.time = mTime
            schedule.year = mCurrentSelectYear
            schedule.month = mCurrentSelectMonth
            schedule.day = mCurrentSelectDay
            AddScheduleTask(mActivity, object : OnTaskFinishedListener<Schedule> {
                override fun onTaskFinished(data: Schedule) {
                    if (data != null) {
                        mScheduleAdapter!!.insertItem(data)
                        etInputContent!!.text.clear()
                        rlNoTask!!.visibility = View.GONE
                        mTime = 0
                        updateTaskHintUi(mScheduleAdapter!!.itemCount - 2)
                    }
                }
            }, schedule).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR)
        }
    }

    private fun setCurrentSelectDate(year: Int, month: Int, day: Int) {
        mCurrentSelectYear = year
        mCurrentSelectMonth = month
        mCurrentSelectDay = day
    }

    override fun onTaskFinished(data: List<Schedule>) {
        mScheduleAdapter!!.changeAllData(data)
        rlNoTask!!.visibility = if (data.isEmpty()) View.VISIBLE else View.GONE
        updateTaskHintUi(data.size)
    }

    private fun updateTaskHintUi(size: Int) {
        if (size == 0) {
            slSchedule!!.removeTaskHint(mCurrentSelectDay)
        } else {
            slSchedule!!.addTaskHint(mCurrentSelectDay)
        }
    }

    override fun onSelectDate(year: Int, month: Int, day: Int, time: Long, position: Int) {
        slSchedule!!.monthCalendar.currentItem = position
        slSchedule!!.postDelayed({ slSchedule!!.monthCalendar.currentMonthView.clickThisMonth(year, month, day) }, 100)
        mTime = time
    }

    companion object {

        val instance: ScheduleFragment
            get() = ScheduleFragment()
    }}

  /*  fun swiped(){
        var swipeHelper = SwipeHelper(context, rv_list) {
      @Override
    fun instantiateUnderlayButton(RecyclerView.ViewHolder viewHolder, List<SwipeHelper.UnderlayButton> underlayButtons) {
        underlayButtons.add(SwipeHelper.UnderlayButton(
                "Delete",
                0,
                Color.parseColor("#FF3C30"),
                SwipeHelper.UnderlayButtonClickListener() {
                    override fun onClick(int pos) {
                        // TODO: onDelete
                    }
                }
        ))

        underlayButtons.add(object :SwipeHelper.UnderlayButton(
                "Transfer",
                0,
                Color.parseColor("#FF9502"),
                object :SwipeHelper.UnderlayButtonClickListener() {
                    override fun onClick(pos: Int) {
                        // TODO: OnTransfer
                    }
                }
        ))
        underlayButtons.add(object :SwipeHelper.UnderlayButton(
                "Unshare",
                0,
                Color.parseColor("#C7C7CB"),
                object :SwipeHelper.UnderlayButtonClickListener() {
                    override fun onClick(pos: Int) {
                        // TODO: OnUnshare
                    }
                }
        ))
    }
}
    }
}*/
